src/index.d.ts

```diff
@@ -54,10 +54,12 @@ interface JsonRpcMiddleware {
     next: JsonRpcEngineNextCallback,
     end: JsonRpcEngineEndCallback,
   ) : void;
+  destroy?: () => void;
```

Comment:
What does this function _do_? There also isn't really a equivalent _create_ function so this seems misplaced.

Reply:
I understand that it's inspired by the extension's usage of a similar pattern but codifying it here seems to be accepting that this is the correct place for it, which I don't know that it is?

Reply:
The `create` equivalent is the initialization of the function. Per [the motivating example from the extension](https://github.com/MetaMask/metamask-extension/blob/890bc25e2863162eb27b4c903357212c222ec904/app/scripts/metamask-controller.js#L1566-L1570), sometimes middlewares need cleanup/teardown. If so, pop a `middleware.destroy = () => { ... }` into your code and you're good to go.

For example, see:
- https://github.com/MetaMask/eth-json-rpc-filters/blob/master/index.js/#L63-L66

`teardown` may be a better name, although `destroy` is what we're already using in practice.

Reply:
Didn't see the second message until after I had already responded, hang on a second.

Reply:
I think this is the correct place for the `destroy` function, because otherwise the consumer needs to maintain references to middlewares they may not care about after creating, _and_ figure out when the RPC engine is torn down, in order to tear down its middlewares. It seems safer to handle it in this module's own teardown logic.

Reply:
I guess my concern here is that this type isn't intuitive.

This is a function that has a function on it. That's not _invalid_ per-se (e.g. `apply` or `bind`, but those are quite different) though it is _uncommon_. I think this type would make more sense if we passed in an _object_ with explicit lifecycle methods, for example:

```js
engine.push({
  start() {
  }
  run () {
  }
  stop() {
  }
})
```

But that would require a bit more work and actually defining a proper lifecycle.

Reply:
The middlewares are no longer of concern to this PR.